consensus/istanbul/celo-ibft-spec.md

```diff
@@ -0,0 +1,394 @@
+# Specification for Celo IBFT protocol
+
+## Celo IBFT specification
+
+### High level overview
+
+The Celo IBFT protocol is a BFT (Byzantine Fault Tolerant) protocol that allows
+a group of participants to agree on the ordering of values by exchanging voting
+messages across a network. As long as less than 1/3rd of participants deviate
+from the protocol then the protocol should ensure that there is only one
+ordering of values that participants agree on and also that and that they can
+continue to agree on new values, i.e. they don't get stuck.
+
+Ensuring that only one ordering is agreed upon is referred to as 'safety' and
+ensuring that the participants can continue to agree on new values is referred
+to as 'liveness'.
+
+### System model
+
+We consider a group of `3f+1` participating computers (participants) that are
+able to communicate across a network. The group should be able to tolerate `f`
+failures without losing liveness or safety.
+
+All participants know the other participants in the system and it is assumed
+that all messages are signed in some way such that participants in the protocol
+know which participant a message came from and that messages cannot be forged.
+
+Only messages from participants are considered.
+
+For messages containing a value only messages with valid values are considered.
+
+Participants are able to broadcast messages to all other participants. In the
+case that a participant is off-line or somehow inaccessible they will not
+receive broadcast messages and there is no mechanism for these messages to be
+re-sent.
+
+We refer to a consensus instance to mean consensus for a specific height,
+consensus instances are independent (they have no shared state).
+
+#### Instance state
+`Sc - current participant state`\
+`Hc - current height`\
+`Rc - current round`\
+`Rd - desired round`\
+`Vc - currently proposed value`\
+`PCc - current prepared certificate`\
+
+### Algorithm
+See supporting [functions](#Functions) and [notation](#Appendix-1-Notation).
+```
+upon: FC_E
```

Comment:
I don't quite get this section. Are these pseudocodes for something like an event-driven transition in the node state?

Reply:
Did you see the section describing upon conditions?
https://github.com/celo-org/celo-blockchain/blob/58804ca7236f60b19ed7fd3ec9d0fc3284621b00/consensus/istanbul/celo-ibft-spec.md?plain=1#L328-L339

If that is not making sense let me know.